cohort-util/src/main/java/com/ibm/cohort/valueset/ValueSetUtil.java

```diff
@@ -97,30 +114,43 @@ else if (inCodesSection) {
 		valueSet.setCompose(compose);
 		ValueSetArtifact artifact = new ValueSetArtifact();
 		artifact.setName(valueSet.getName());
-		artifact.setResource(valueSet);
+		artifact.setFhirResource(valueSet);
 		artifact.setUrl(valueSet.getUrl());
 		return artifact;
 	}
 
-	public static void importArtifacts(IGenericClient client, List<ValueSetArtifact> valueSetArtifacts) {
-		importArtifacts(client, valueSetArtifacts, false);
+	public static void importArtifacts(IGenericClient client, List<ValueSetArtifact> valueSetArtifacts, boolean updateIfExists) {
+		for(ValueSetArtifact valueSetArtifact : valueSetArtifacts) {
+			importArtifact(client, valueSetArtifact, updateIfExists);
+		}
+	}
+
+	public static void deleteValueSet(IGenericClient client, String url){
+		client.delete().resourceConditionalByType("ValueSet").where(ValueSet.URL.matches().value(url)).execute();
 	}
 
-	public static void importArtifacts(IGenericClient client, List<ValueSetArtifact> valueSetArtifacts, boolean continueIFExists) {
+	public static String importArtifact(IGenericClient client, ValueSetArtifact valueSetArtifact, boolean updateIfExists) {
 
-		for (ValueSetArtifact valueSetArtifact : valueSetArtifacts) {
-			Bundle bundle = client.search().forResource(ValueSet.class).where(ValueSet.URL.matches().value(valueSetArtifact.getUrl()))
-					.returnBundle(Bundle.class).execute();
-			if (bundle.getEntry().size() > 0) {
-				valueSetArtifact.setId(bundle.getEntryFirstRep().getResource().getIdElement().getIdPart());
-				if (continueIFExists)  {
-					MethodOutcome outcome = client.update().resource(client.getFhirContext().newJsonParser().encodeResourceToString(valueSetArtifact.getResource())).execute();
-					valueSetArtifact.setId(outcome.getId().getIdPart());
-				}
-			} else {
-				MethodOutcome outcome = client.create().resource(client.getFhirContext().newJsonParser().encodeResourceToString(valueSetArtifact.getResource())).execute();
-				valueSetArtifact.setId(outcome.getId().getIdPart());
+		Bundle bundle = client.search().forResource(ValueSet.class).where(ValueSet.URL.matches().value(valueSetArtifact.getUrl()))
+				.returnBundle(Bundle.class).execute();
+		MethodOutcome outcome;
+		if(bundle.getEntry().size() > 0){
+			String[] url = bundle.getEntry().get(0).getFullUrl().split("/");
+			valueSetArtifact.getFhirResource().setId(url[url.length-1]);
+			if(updateIfExists){
+				outcome = client.update().resource(client.getFhirContext().newJsonParser().encodeResourceToString(valueSetArtifact.getFhirResource()))
+						.conditional().where(ValueSet.URL.matches().value(valueSetArtifact.getUrl())).execute();
+			}
+			else{
+				return null;
 			}
 		}
+		else {
+			outcome = client.create().resource(client.getFhirContext().newJsonParser().encodeResourceToString(valueSetArtifact.getFhirResource())).execute();
+		}
```

Comment:
There is a race condition here between when the search method returns and the create/update method runs. During execution, we might try to create a resource that shows up after the search completes or update a resource that has changed since we last saw it.  FHIR has [capabilities](https://www.hl7.org/fhir/http.html#concurrency) built in to handle conditional creates and concurrency issues and there are [HAPI client examples](https://hapifhir.io/hapi-fhir/docs/client/examples.html) available for how to use them. It looks something like this...

``` java
bundle.addEntry()
   .setFullUrl(patient.getIdElement().getValue())
   .setResource(patient)
   .getRequest()
      .setUrl("Patient")
      .setIfNoneExist("identifier=http://acme.org/mrns|12345")
      .setMethod(Bundle.HTTPVerb.POST);
```

The odds here of the race condition mattering are pretty slim. I am *not* requesting changes to your PR. I just want to document this because it is something I thought about when I created the MeasureImporter. In the end, I just did a PUT for everything in my bundle because PUT in IBM FHIR will do [update as create](https://www.hl7.org/fhir/http.html#upsert) if the resource isn't there (as long as you provide a Resource.id in your resource). I didn't have a requirement for the "block if exists" behavior.

Reply:
Thanks for this, definitely has given me something to think about.